New data modeling tutorial (#437)
neo4j/docs-getting-started: modules/ROOT/pages/data-modeling/tutorial-data-modeling.adoc

= Tutorial: Create a graph data model
:tags: graph-modeling, data-model, schema, model-process, model-domain, tutorial
:description: This tutorial is designed to help you understand how data modeling works through an example case.

> This tutorial is designed to help you understand how data modeling works through an example case.

Reviewer: ```suggestion
This tutorial is designed to help you understand how to model your data based on what you intend to use it for.
```

Author: Not really what you intend to use it for, because it uses a specific use case as an example, and not a general one, one size fits all, sort of.

Reviewer: No but the idea is general, that you model the data based on what you intend to use it for. In this case we want to use the example data to answer some questions and model it based on those. Do you see what I mean?

Author: yeah, but I don't know if I agree to put it like this here. maybe a middle term could be like

```suggestion
This tutorial is designed to help you understand how to model your data based on what you intend to use it for, using an example case.
```

> You will use the xref:appendix/example-data.adoc[Movies example dataset] as the main resource.

Reviewer 2: Duplicating what is stated in line 15 below

[TIP]
====
For an interactive course on the fundamentals of data modeling, see link:https://graphacademy.neo4j.com/courses/modeling-fundamentals/?ref=docs[GraphAcademy].
====

== Define the domain

> In this tutorial, you will use the xref:appendix/example-data.adoc[Movies example dataset], which means your domain includes movies, people who acted or directed movies, and users who rated movies. 

Reviewer 3: ```suggestion
In this tutorial, you will use the xref:appendix/example-data.adoc[Movies example dataset], the domain includes movies, people who acted or directed movies, and users who rated movies. 
```

> This helps you to have a better idea of what the application will be, which people and systems will use it, and thus which domain you will work with.

Author: This is what is written in the GraphAcademy course https://graphacademy.neo4j.com/courses/modeling-fundamentals/1-getting-started/2-the-domain/

> This helps you to have a better idea of what the application will be, which people and systems will use it, and thus which domain you will work with.

Reviewer: Yes, but I'm not sure it is very helpful. But I'll leave that up to you to decide.

== Define the use case

> With the domain defined, you need to identify what the use case is for your application.

Reviewer 3: ```suggestion
With the domain defined, you need to identify your application use cases.
```

In other words, what questions are you trying to answer?

> You can make a *list of questions* to help you identify the use cases of your application.

Reviewer 3: ```suggestion
You can make a *list of questions* to help you identify the application use cases.
```

> It is in the connections (relationships) between these entities that you find interesting information about your domain.

Author: Also taken from GA, but I suppose this sentence can be tweaked to something like 

```suggestion
It is in the connections (relationships) between these entities that you find insights about your domain.
```

Or remove it.

> It is in the connections (relationships) between these entities that you find interesting information about your domain.

Reviewer: That's better, I think.

> This way, you can better define what you need from it, and what data must be included in the graph.

Reviewer 3: ```suggestion
The questions will help you define what you need from the application, and what data must be included in the graph.
```

> For this tutorial, consider these the questions that you want answered by your application:

Reviewer 3: ```suggestion
For this tutorial, your application should be able to answer these questions:
```

> With this information, you can better define how the application will behave at runtime.

Author: This is recommendation from the GA course too.

> With this information, you can better define how the application will behave at runtime.

Reviewer: Maybe we should try to differentiate this one a little? If we don't give any further explanation to this, I think it is better to omit altogether.

> Most use cases for an application can be defined with a *list of questions*.
> With this information, you can better define how the application will behave at runtime.

Author: But I think it's good advice, so you can understand what kind of data you need to add to the graph in order to make it functional for the use cases. I can tweak the sentences a bit, maybe something like:

```suggestion
You can make a *list of questions* to help you identify the use cases of your application.
This way, you can better define what you need from it, and what data needs to be added to the graph.
```

> With this information, you can better define how the application will behave at runtime.

Reviewer: Or maybe ".., and what data must be included in the graph". At this point, there is no data in the graph.

For this tutorial, consider these as the questions that need to be answered by the application:

* What people acted in a movie?
* What person directed a movie?
* What movies did a person act in?
* How many users rated a movie?
* Who was the youngest person to act in a movie?
* What role did a person play in a movie?
* What is the highest rated movie in a particular year according to imDB?
* What drama movies did an actor act in?

> * What users gave a movie a rating of 5?

Author: No, but in the testing phase, described more ahead, as you realize there are remaining use cases to be answered.

image::sample-data-model.svg[Sample data model featuring two nodes for person and movie connected through two relationships, one for directed and another for acted in,role=popup]

Note that in the model, labels, relationship types, and property keys follow a certain syntax.
In Cypher, these are called identifiers and they are case-sensitive, as are string values.

See the Cypher link:{docs-home}/cypher-manual/current/styleguide[style guide] for more information.
While not mandatory, it is recommended that:

* Labels are capitalized and can be CamelCase (e.g., `Person`, `Movie`, `ImdbUser`).
* Relationship types are written with all capital letters and an underscore character as a separator (e.g., `DIRECTED`, `ACTED_IN`).
* Property keys for nodes or relationships are *not* capitalized and can be camelCase (e.g. `name`, `userID`).

[NOTE]
====
At this stage of creating your initial model, focus on the high-level design of your model, i.e. how your entities connect.
The xref:#_define_entities[Define entities] step describes a more detailed view on how to allocate certain information in a graph (e.g., as a node, relationship, property, etc).
====

=== Instance model

> 

Reviewer 3: Can you include a definition of what an instance model is (in the same way you do for a domain model?)

=== Instance model

An important part of the graph data modeling process is to test the model against the use cases.
For that, you need to have a set of sample data that you can use to check that:

image::sample-data-instance-model.svg[Sample data model contaning nodes for movies and people with the properties title, name, date of release, birth, rating, and ID,500,500,role=popup]

> 

Author: I will add an additional step here to create an instance in the user's preferred deployment.

CREATE (Apollo13:Movie {title: 'Apollo 13', tmdbID: 568, released: '1995-06-30', imdbRating: 7.6, genres: ['Drama', 'Adventure', 'IMAX']})
CREATE (TomH:Person {name: 'Tom Hanks', tmdbID: 31, born: '1956-07-09'})
CREATE (MegR:Person {name: 'Meg Ryan', tmdbID: 5344, born: '1961-11-19'})
CREATE (DannyD:Person {name: 'Danny DeVito', tmdbID: 518, born: '1944-11-17'})
CREATE (JackN:Person {name: 'Jack Nicholson', tmdbID: 514, born: '1937-04-22'})
CREATE (SleeplessInSeattle:Movie {title: 'Sleepless in Seattle', tmdbID: 858, released: '1993-06-25', imdbRating: 6.8, genres: ['Comedy', 'Drama', 'Romance']})
CREATE (Hoffa:Movie {title: 'Hoffa', tmdbID: 10410, released: '1992-12-25', imdbRating: 6.6, genres: ['Crime', 'Drama']})
--

[NOTE]
====
The data used here can be found in the xref:appendix/example-data.adoc[Movies example dataset], also available in Browser and Aura guides.
However, in order to practice data modeling, it is recommended that you add the data manually using Cypher.
====

== Define entities

> Creating an instance model in the previous step helps you preview how you may define the data as nodes, relationships, and properties.

Reviewer 3: ```suggestion
An instance model helps you preview how the data will be stored as nodes, relationships, and properties.
```

The next step is to refine your model with more details.

=== Labels

> The dominant nouns in your application use case are represented as nodes in your model and are used as *node labels*.

Reviewer 3: ```suggestion
The dominant nouns in your application use case are represented as nodes in your model and can be used as *node labels*.
```

> 

Author: Yes, but this is intended, for when you test the model you realize that you missed something.

> However, in order to practice data modeling, it is recommended that you add the data manually using Cypher or link:{docs-home}/data-importer/modeling[Data Importer].

Author: Yeah, better to remove it. Thanks for catching it!

> == Define entities

Author: It's how it is called in the GA course, and I think it means more like double checking if what you set as a node/relationship before would still work as that or as a property, for example, at this stage.

Entities are the dominant nouns in your application use cases.
For example:

* What [.underline]#person# acted in a [.underline]#movie#?

> * How many [.underline]#users# rated a [.underline]#movie#?

Author: True, but then we would need to separate actors and directors as different kinds of nodes than Person or add user information in Person nodes. I added a note more below about that. Maybe readers don't pay so much attention to that if we have some explanation further into the tutorial?

Although this could change after xref:data-modeling/graph-model-refactoring.adoc[refactoring], you can assume that the dominant nouns "person", "movie", and "user" will be nodes in the first iteration of the data model.

=== Node properties

> Node properties are used to uniquely identify a node, answer specific details of the use cases for the application, and to return data.

Author: When I need to know the release date of a movie, and query the property m.release, for example.

> Node properties are used to uniquely identify a node, answer specific details of the use cases for the application, and to return data.

Reviewer: The query returns the data, not the property. Maybe this can be rephrased to make that a little clearer?

For example, in a Cypher statement, properties can be used to:

.Anchor (where to begin the query)
[source,cypher]
--
MATCH (p:Person {name: 'Tom Hanks'})-[:ACTED_IN]-(m:Movie) RETURN m
--

.Traverse the graph (navigation)
[source,cypher]
--
MATCH (p:Person)-[:ACTED_IN]-(m:Movie {title: 'Apollo 13'})-[:RATED]-(u:User) RETURN p,u
--

> .Return data from the query.

Author: But you query the property to get the data from the property value, no?

.Return data from the query
[source,cypher]
--
MATCH (p:Person {name: 'Tom Hanks'})-[:ACTED_IN]-(m:Movie) 
RETURN m.title, m.released
--

Since you are interested in how individual people are related to individual movies, you want each instance of an entity (each different person and each different movie) to be a separate node.
In other words, every instance of your model's *Person* node is a distinct node and you use properties to separate them.
For example, the `Person` node with the property `name` and the value 'Tom Hanks' is distinct from the `Person` node with the property value `Meg Ryan` for the same property. 

==== Unique identifiers

You can use properties to uniquely identify nodes.

In this tutorial, you are working with movies and people, both entities that have names.

> While you can create a property called `name` for both `Person` and `Movie` nodes, it is advisable to use different terms to provide unique identifiers.

Reviewer 3: Is this true? I dont see why it would be important to have different unique property name for different node labels.

Reviewer 3: I think its more important to have property names which reflect the field.

--

| What movies did a person act in?
| * Retrieve a person by their *name*.
* Return the *titles* of the movies.
|[source,cypher]
--
MATCH (p:Person {name:'Tom Hanks'})-[:ACTED_IN]->(m:Movie)
RETURN m.title
--

| Who was the youngest person to act in a movie?
| * Retrieve a movie by its *title*.
* Evaluate the *ages* of the actors.
* Return the *name* of the actor.
|[source,cypher]
--
MATCH (m:Movie {title:'Sleepless in Seattle'})<-[r:ACTED_IN]-(p:Person)
RETURN p.name, p.born
ORDER BY p.born ASC
LIMIT 1
--

| What is the highest rated movie in a particular year according to imDB?
| * Retrieve all movies *released* in a particular year.
* Evaluate the *imDB ratings*.
* Return the movie *title*.
|[source,cypher]
--
MATCH (m:Movie {release:date('1995')})
RETURN m.title, m.imdbRating
ORDER BY m.imdbRating DESC
LIMIT 1
--

|===

=== Relationships

Relationships are connections between entities and these connections are the verbs in your use cases:

* What person [.underline]#acted in# a movie?
* What person [.underline]#directed# a movie?

At a glance, connections seem straightforward, but their micro- and macro-design are arguably the most critical factors in graph performance. 

> To get started, thinking of relationships from the perspective that “connections are verbs” works well, but there are other important considerations that you will learn as you advance with your model.

Author: I think it sounds weird without the 'will'.

At a glance, connections seem straightforward, but their micro- and macro-design are arguably the most critical factors in graph performance. 
To get started, thinking of relationships from the perspective that “connections are verbs” works well, but there are other important considerations that you will learn as you advance with your model.

==== Naming

It is important to choose good names (types) for the relationships in the graph.
Relationship types need to be intuitive and not confused with node labels.

For the example use cases, you could define relationships as:

* `ACTED_IN`
* `DIRECTED`

With these options, you can already plan the direction of the relationships.

==== Relationship direction

All relationships must have a direction.
When created, relationships need to specify their direction explicity or inferred by the left-to-right order of the pattern.

> When you query the database, however, you don't need to specify the direction, as long as the direction is inferred by the order of the pattern in a query.

Reviewer 3: Minor point, but Im not sure this is important for data modelling. It may just add confusion.

To add all `ACTED_IN` and `DIRECTED` relationships, you can run this code:

[source,cypher]
--
MERGE (TomH)-[:ACTED_IN]->(Apollo13)
MERGE (TomH)-[:ACTED_IN]->(SleeplessInSeattle)
MERGE (MegR)-[:ACTED_IN]->(SleeplessInSeattle)
MERGE (DannyD)-[:ACTED_IN]->(Hoffa)
MERGE (DannyD)-[:DIRECTED]->(Hoffa)
MERGE (JackN)-[:ACTED_IN]->(Hoffa)
--

And your graph should now look like this:

image::relationships-graph.svg[All person nodes are now connected to the movie nodes through an acted in or directed relationship,role=popup]

[TIP]
====
You can always use the query `MATCH (n) RETURN n` to check how your graph looks like.
====

==== Super nodes

> Now, consider a hypothetical scenario in which you decided to have the actors as separate nodes rather than referring to them using the `name` property in `Person` nodes:

Author: The actors are Person nodes and their names/differentiation are found in the property 'name'. When created, they were configured like this:

CREATE (TomH:Person {name:'Tom Hanks'})
CREATE (MegR:Person {name:'Meg Ryan'})

The example here assumes that each actor is a single different role, and so created like this:

CREATE (t:TomH {name:'Tom Hanks'})
CREATE (m:MegR {name:'Meg Ryan'})

But the image is misleading, the node label should be like TomH instead of Tom Hanks, for example. I need to fix that.

> Now, consider a hypothetical scenario in which you decided to have the actors as separate nodes rather than referring to them using the `name` property in `Person` nodes:

Reviewer: I see what you mean, but the image is not really doing us any favors here. It would look very similar with the Person label.
What you mean is that the actor nodes do not have a Person label? Even with the label to unify the Person nodes, the movie node will still be a super node.

> Now, consider a hypothetical scenario in which you decided to have the actors as separate nodes rather than referring to them using the `name` property in `Person` nodes:

Author: I just gave it a second thought and I think this topic is too complex to be addressed here. Alternatively, we could mention the case of super nodes and link to this developer blog article instead -- https://medium.com/neo4j/graph-modeling-all-about-super-nodes-d6ad7e11015b . What do you think?

==== Relationship properties

Properties for a relationship are used to enrich how two nodes are related. 
When you need to know *how* two nodes are related and not just that they are related, you can use relationship properties to further define the relationship.

The example question "Which role did a person play in a movie?" can be asked with the help of the property `roles` in the `ACTED_IN` relationship:

image::relationship-roles.svg[Example showing how a person node connected to a movie node with the relationship acted in which has the property role,400,400,role=popup]

Note that the information about roles needs to be added to the graph before being retrieved.
You can use this Cypher statement for that:

[source,cypher]
--
MERGE (TomH)-[:ACTED_IN {roles:'Jim Lovell'}]->(Apollo13)
MERGE (TomH)-[:ACTED_IN {roles:'Sam Baldwin'}]->(SleeplessInSeattle)
MERGE (MegR)-[:ACTED_IN {roles:'Annie Reed'}]->(SleeplessInSeattle)
MERGE (DannyD)-[:ACTED_IN {roles:'Robert "Bobby" Ciaro'}]->(Hoffa)
MERGE (JackN)-[:ACTED_IN {roles:'Hoffa'}]->(Hoffa)
--

Then, in order to find which role Tom Hanks played in Apollo 13, you use the following statement:

[source,cypher]
--

> MATCH (TomH)-[r:ACTED_IN]->(Apollo13)

Reviewer 3: This cypher wouldnt run independently (i.e. without being used at the end of the previous Cypher). The variables `TomH` and `Apollo13` arent defined.

Author: Well then all other previous code examples using the variables TomH etc, will need to be updated too, right?

Reviewer 3: I think its just this one.

Author: I updated it. Is it correct now or does it still need the MERGE lines like you suggested?

Reviewer 3: Looks good to me 😄

Author: Awesome, so all your suggestions were already applied in case you want to take a last look 😁

> In the example, the use case "What role did a person play in a movie?" could be asked with the help of the property `roles` in the `ACTED_IN` relationship:

Reviewer: ```suggestion
The example question "What role did a person play in a movie?" can be asked with the help of the property `roles` in the `ACTED_IN` relationship:
```

> Note that the information about roles needs to be added to the graph before being retrieved:

Reviewer: ```suggestion
You can use this Cypher statement to add information about roles to the `ACTED_IN`relationship:
```

--

Then, in order to check what role Tom Hanks played in Apollo 13, you use the following code:

[source,cypher]
--
MATCH (TomH)-[r:ACTED_IN]->(Apollo13)
RETURN r.roles
--

With the addition of the new relationship property, your graph should now look like this:

image::roles-graph.svg[Graph with person and movies nodes, and acted in relationships with the property roles, 400,400,role=popup]

== Add more data

Now that you have created the first connections between the nodes, it's time to add more information to the graph.

> This way, you can work with other use cases, such as:

Reviewer: ```suggestion
This way, you can answer more questions, such as:
```

* How many users rated a movie?

> * What users gave a movie a rating of 5?

Author: ```suggestion
* Which users gave a movie a rating of 5?
```

> To answer these questions, you need information about users and their ratings in your graph, which means a change in your data model.

Reviewer: This would be a good spot to add something about users also being persons and the various ways we can deal with this in the data model.

Note that, with the addition of new data such as the property `roles` in the `ACTED_IN` relationship, your xref:#_data_model[initial data model] has already been updated along the way:

image::updated-model.svg[Addition of a new user node to the graph,role=popup]

> You can start first by adding the users to your graph:

Reviewer: ```suggestion
You can start by adding the users to your graph:
```

====

Then, connect the `User` nodes to the `Movie` nodes through a `RATED` relationship which contains the `rating` property:

[source,cypher]
--
MERGE (Sandy)-[:RATED {rating:5}]->(Apollo13)
MERGE (Sandy)-[:RATED {rating:4}]->(SleeplessInSeattle)
MERGE (Clinton)-[:RATED {rating:3}]->(Apollo13)
MERGE (Clinton)-[:RATED {rating:3}]->(SleeplessInSeattle)
MERGE (Clinton)-[:RATED {rating:3}]->(Hoffa)
--

Your graph should now look like this:

image::user-ratings.svg["Graph containing person, movie, and user nodes connected through acted in, directed, and rated relationships", 500,500,role=popup]

== Test the model

After populating the graph to implement the data model with a small set of test data, you should now test it to ensure that it satisfies every xref:#_define_the_use_case[use case].

For example, if you want to test the use case "Which people acted in a movie?", you can run the following query:

[source,cypher]
--
MATCH (p:Person)-[:ACTED_IN]-(m:Movie)
WHERE m.title = 'Sleepless in Seattle'
RETURN p.name
--

This is just a simple example of testing.

> As you go through the use cases, you may think of more data to be added to the graph in order to round out the testing.

Reviewer 3: ```suggestion
As you go through the use cases, you may think of more data to be added to the graph in order to complete the testing.
```

Additionally, make sure that the Cypher statements used to test the use cases are correct.
A query written incorrectly could lead to the assumption that the data model has failed.

> For example, if you want want to find a user, but you forgot that their data is stored as a `User` node rather than a `Person` node, when you query for them using the `Person` node, you won't get any information and assume that they don't exist in the graph.

Reviewer 3: ```suggestion
For example, using an incorrect node label in a test may lead you to believe that the data doesn't exist in the graph.
```

> For example, if you want to test the use case "What people acted in a movie?", you can execute the following query:

Reviewer: ```suggestion
For example, if you want to test your model with the question "What people acted in a movie?", you can run the following query:
```

> As you go through the use cases, you may think of more data that you want to add to the graph to round out the testing.

Reviewer: ```suggestion
As you go through your question, you may think of more data to add to the graph to round out the testing.
```

> Additionally, make sure that the Cypher code used to test the use cases is correct.

Reviewer: ```suggestion
Additionally, make sure that the Cypher statements used to test the use cases are correct.
```

> A query written incorrectly could lead to the assumption that the data model has failed, for example.

Author: Yes, or maybe, for instance, I want to find a user and I forget they are a User node rather than a Person node, so when I query for them using the Person node, then I might think that they don't exist.

> A query written incorrectly could lead to the assumption that the data model has failed, for example.

Reviewer: I see! I think we should have an example to clarify that.

> At this point, you will also start considering scalability of your graph and how performant it would be if you write the same queries in a graph with millions of nodes and relationships.

Reviewer: ```suggestion
At this point, you can also start considering the scalability of your graph and how performant it would be if you write the same queries in a graph with millions of nodes and relationships.
```

|[source,cypher]
--
MATCH (m:Movie {title:'Sleepless in Seattle'})<-[r:ACTED_IN]-(p:Person)
RETURN p.name, p.born
ORDER BY p.born ASC
LIMIT 1
--

| What is the highest rated movie in a particular year according to imDB?
| * Retrieve all movies *released* in a particular year.
* Evaluate the *imDB ratings*.
* Return the movie *title* for the movie with the highest rating.
|[source,cypher]
--
MATCH (m:Movie {release:date('1995')})
RETURN m.title, m.imdbRating
ORDER BY m.imdbRating DESC
LIMIT 1
--

|===

=== Relationships

Relationships are connections between nodes, and these connections are the verbs in your use cases:

* Which person [.underline]#acted in# a movie?
* Which person [.underline]#directed# a movie?

At a glance, connections seem straightforward, but their micro- and macro-design are arguably the most critical factors in graph performance. 
To get started, thinking of relationships from the perspective that “connections are verbs” works well, but there are other important considerations that you will learn as you advance with your model.

==== Naming

It is important to choose good names (types) for the relationships in the graph.

> Relationship types need to be intuitive and not confused with node labels.

Reviewer 4: Relationship types need to be specific too.
I thing I often notice when reviewing data models is vague naming such as `HAS`, `RELATED`, `CONNECTED_TO`... As a consequence, the queries are not very expressive. This might cause performance problems as well for dense nodes: Neo4j optimizes queries based on relationship types. If all relationships have the same name, queries must scan many more relationships, slowing down performance. Whereas More specific relationships allow Neo4j to traverse only relevant connections

MERGE (Clinton)-[:RATED {rating:3}]->(SleeplessInSeattle)
MERGE (Clinton)-[:RATED {rating:3}]->(Hoffa)
--

Your graph should now look like this:

image::user-ratings.svg["Graph containing person, movie, and user nodes connected through acted in, directed, and rated relationships", 500,500,role=popup]

== Test the model

After populating the graph to implement the data model with a small set of test data, you should now test it to ensure that it satisfies every xref:#_define_the_use_case[use case].

For example, if you want to test the use case "Which people acted in a movie?", you can run the following query:

[source,cypher]
--
MATCH (p:Person)-[:ACTED_IN]-(m:Movie)
WHERE m.title = 'Sleepless in Seattle'
RETURN p.name
--

This is just a simple example of testing.
As you go through the use cases, you may think of more data to be added to the graph in order to complete the testing.

Additionally, make sure that the Cypher statements used to test the use cases are correct.
A query written incorrectly could lead to the assumption that the data model has failed.

For example, using an incorrect node label in a test may lead you to believe that the data doesn't exist in the graph.

> At this point, you can also start considering the scalability of your graph and how performant it would be if you write the same queries in a graph with millions of nodes and relationships.

Reviewer 4: how? maybe point the the profiling docs?

| * Retrieve all movies *released* in a particular year.
* Evaluate the *imDB ratings*.
* Return the movie *title* for the movie with the highest rating.
|[source,cypher]
--
MATCH (m:Movie {release:date('1995')})
RETURN m.title, m.imdbRating
ORDER BY m.imdbRating DESC
LIMIT 1
--

|===

==== Unique identifiers

In Cypher, it is possible to create two different nodes with the exact same data.
However, from a data management and model perspective, different nodes should contain different data.
You can use *unique identifiers* to make sure that every node is a separate and distinguished entity.

In the initial instance model, these are the properties set for the `Movies` nodes:

* `Movie.title` (string)
* `Movie.tmdbID` (integer)
* `Movie.released` (date)
* `Movie.imdbRating` (decimal between 0-10)
* `Movie.genres` (list of strings)

And for the `Person` nodes:

* `Person.name` (string)
* `Person.tmdbID` (integer)
* `Person.born` (date)

The `Movie` node property `tmdbID` is a good example of a unique identifier, as there might be different movies with the same title in the database, but the property will be different and thus function as a unique identifier.

[TIP]
====

> You can enforce unique identifiers by using uniqueness constrains.

Reviewer 4: suggestion: replace this with "It is strongly suggested that you enforce..."
